src/distributions/bernoulli.rs

```diff
@@ -0,0 +1,113 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// https://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+//! The Bernoulli distribution.
+
+use Rng;
+use distributions::Distribution;
+
+/// The Bernoulli distribution.
+///
+/// This is a special case of the Binomial distribution where `n = 1`.
+///
+/// # Example
+///
+/// ```rust
+/// use rand::distributions::{Bernoulli, Distribution};
+///
+/// let d = Bernoulli::new(0.3);
+/// let v = d.sample(&mut rand::thread_rng());
+/// println!("{} is from a Bernoulli distribution", v);
+/// ```
+///
+/// # Precision
+///
+/// This `Bernoulli` distribution uses 64 bits from the RNG (a `u64`), making
+/// its bias less than 1 in 2<sup>64</sup>. In practice the floating point
+/// accuracy of `p` will usually be the limiting factor.
+#[derive(Clone, Copy, Debug)]
+pub struct Bernoulli {
+    /// Probability of success, relative to the maximal integer.
+    p_int: u64,
+}
+
+impl Bernoulli {
+    /// Construct a new `Bernoulli` with the given probability of success `p`.
+    ///
+    /// # Panics
+    ///
+    /// If `p < 0` or `p > 1`.
+    ///
+    /// # Precision
+    ///
+    /// For `p = 1.0`, the resulting distribution will always generate true.
+    /// For `p = 0.0`, the resulting distribution will always generate false.
+    #[inline]
+    pub fn new(p: f64) -> Bernoulli {
+        assert!(p >= 0.0, "Bernoulli::new called with p < 0");
+        assert!(p <= 1.0, "Bernoulli::new called with p > 1");
+        let p_int = if p < 1.0 {
```

Comment:
Here we could allow `p > 1` very easily — should we? My fear is that accumulated floating point errors could push a number slightly above 1 or below 0 when it shouldn't be. Unfortunately we cannot support <0 here without a jump.

Alternatively we could use strict bounds here but suggest use of `p > rng.sample(HighPrecision01)` or similar where strict bounds may be a problem.

Reply:
Yes, I thought about this. It would reduce the number of branches, but might mask errors. Accumulated FP errors will happen, but usually the user has to fix those anyway. Adding a branch for `p < 0` is not a problem if we allow `p < 0`, because then we can get rid of the asserts. Another advantage would be that we can make it panic free. What do you think?

Reply:
I think this makes sense, except for one thing: if `p` is NaN, the method should panic, since NaNs usually indicate serious problems which should be fixed. The same could be said for values not close to the range [0, 1]. Because of this maybe you are right, that users should fix accumulated FP errors themselves.

BTW wouldn't it reduce the number of branches if we used a single `assert!(p >= 0.0 && p <= 1.0, ...)` instead?